### PR TITLE
`decoding_method` argument in generate

### PR DESCRIPTION
**Pull request by manueldeprada**
`custom_generate` already allows a Hub/local repository to override `generate()` with a custom implementation.

This PR extends the feature to also accept **custom decoding functions** directly.
When a function is provided, GenerationMixin.generate still handles all input preparation and configuration, and the function only needs to implement the decoding loop.

This enables moving built-in decoding methods such as DoLa and contrastive decoding to the Hub, and power users to easily plug in their own custom loops.

[Docs preview of the new feature here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_40085/en/generation_strategies#reusing-generate-s-preparation-steps-by-passing-a-callable)

### CODE REVIEW

File: src/transformers/generation/utils.py [hunk 1]
```diff
@@ -2352,6 +2354,22 @@ def generate(
         else:
             input_ids = inputs_tensor if model_input_name == "input_ids" else model_kwargs.pop("input_ids")
 
+        # Expand inputs depending on the generation mode
```
Comment (manueldeprada) [inline src/transformers/generation/utils.py:2367]: we need to factor this out of the GenerationMethod if/elif block so that custom loops can also benefit

File: docs/source/en/generation_strategies.md [hunk 2]
```diff
@@ -504,6 +504,30 @@ Recommended practices:
 - Add self-contained examples to enable quick experimentation.
 - Describe soft-requirements such as if the method only works well with a certain family of models.
 
+### Reusing `generate`’s preparation steps by passing a callable
```
Comment (gante) [inline docs/source/en/generation_strategies.md]:
```suggestion
### Reusing `generate`’s input preparation
```
(cleaner + shorter)

File: docs/source/en/generation_strategies.md [hunk 3]
```diff
@@ -504,6 +504,30 @@ Recommended practices:
...
+
+output = model.generate(
+    **inputs,
+    custom_generate=custom_loop,
+    max_new_tokens=10,
+)
+```
+
+> [!TIP]
+> If you publish a `custom_generate` repository, your `generate` implementation can itself define a callable and pass it to `model.generate()`. This lets you customize the decoding loop while still benefiting from Transformers’ built-in preparation logic.
```
Comment (gante) [inline docs/source/en/generation_strategies.md]:
```suggestion
> If you publish a `custom_generate` repository, your `generate` implementation can itself define a callable and pass it to `model.generate()`. This lets you customize the decoding loop while still benefiting from Transformers’ built-in input preparation logic.
```

File: docs/source/en/generation_strategies.md [hunk 4]
```diff
@@ -504,6 +504,30 @@ Recommended practices:
 - Add self-contained examples to enable quick experimentation.
 - Describe soft-requirements such as if the method only works well with a certain family of models.
 
+### Reusing `generate`’s preparation steps by passing a callable
+
+You can also pass a **callable** to `custom_generate` to reuse [`~GenerationMixin.generate`]’s full preparation pipeline (batch expansion, attention masks, logits processors, stopping criteria, etc.) while overriding only the decoding loop.
```
Comment (gante) [inline docs/source/en/generation_strategies.md]:
```suggestion
If you're adding a new decoding loop, you might want to preserve the input preparation present in `generate` (batch expansion, attention masks, logits processors, stopping criteria, etc.). You can also pass a **callable** to `custom_generate` to reuse [`~GenerationMixin.generate`]’s full preparation pipeline while overriding only the decoding loop.
```
(slightly rearranging to place emphasis on when is this useful)

File: src/transformers/generation/utils.py [hunk 5]
```diff
@@ -2227,11 +2227,13 @@ def generate(
...
-                function defined in that reposity's `custom_generate/generate.py` file will be executed instead of the
-                standard `generate` method. Note that the logic is for generation is entirely defined in that
-                repository, and the return type may be different from the standard `generate` method.
+            custom_generate (`str` or `Callable`, *optional*):
+                A string containing the name of a huggingface.co repository, a local repository path, or a callable.
+                If a repository is provided, the custom `generate` function defined in that repository's
+                `custom_generate/generate.py` file will be executed instead of the standard `generate` method. Note
+                that the logic for generation will be entirely replaced by the repository, and the return type may be
+                different from the standard `generate` method. If a callable is provided, `generate` will still run the
+                standard preparation steps, but will call the provided callable to perform the decoding loop.
```
Comment (gante) [inline src/transformers/generation/utils.py]: two suggestions:
1. let's rearrange the text as a list, where each list entry is an input type. Goal: easier to visually parse what each input type does
2. let's add a link to the docs at the end of this docstring. Goal: quickly redirect to the full information

File: src/transformers/generation/utils.py [hunk 6]
```diff
@@ -2352,6 +2354,22 @@ def generate(
...
+                is_encoder_decoder=self.config.is_encoder_decoder,
+                **model_kwargs,
+            )
+        elif generation_config.do_sample:
+            input_ids, model_kwargs = self._expand_inputs_for_generation(
+                input_ids=input_ids,
+                expand_size=generation_config.num_return_sequences,
+                is_encoder_decoder=self.config.is_encoder_decoder,
+                **model_kwargs,
+            )
```
Comment (gante) [inline src/transformers/generation/utils.py]:
```suggestion
        input_ids, model_kwargs = self._expand_inputs_for_generation(
            input_ids=input_ids,
            expand_size=max(generation_config.num_beams, generation_config.num_return_sequences)
            is_encoder_decoder=self.config.is_encoder_decoder,
            **model_kwargs,
        )
```

we can take advantage of:
- `num_beams` and `num_return_sequences` both defaulting to `1`
- `expand_size=1` is a no-op

to make it shorter :D

File: src/transformers/generation/utils.py [hunk 7]
```diff
@@ -2433,7 +2451,18 @@ def generate(
         model_kwargs["use_cache"] = generation_config.use_cache
 
         # 10. go into different generation modes
-        if generation_mode == GenerationMode.ASSISTED_GENERATION:
+        if custom_generate is not None and isinstance(custom_generate, Callable):
```
Comment (gante) [inline src/transformers/generation/utils.py]:
```suggestion
        if isinstance(custom_generate, Callable):
```
(redundant?)

File: src/transformers/generation/utils.py [hunk 8]
```diff
@@ -2575,14 +2589,7 @@ def generate(
...
             )
-            # 12. interleave input_ids with `num_beams` additional sequences per batch
-            input_ids, model_kwargs = self._expand_inputs_for_generation(
-                input_ids=input_ids,
-                expand_size=generation_config.num_beams,
-                is_encoder_decoder=self.config.is_encoder_decoder,
-                **model_kwargs,
-            )
-            # 13. run beam search
+            final_constraints = []
```
Comment (gante) [inline src/transformers/generation/utils.py]:
```suggestion
```
(?)

File: tests/generation/test_utils.py [hunk 9]
```diff
@@ -5044,6 +5044,26 @@ def test_custom_generate_local_directory(self):
             )
             assert value == "success"
 
+    def test_custom_generate_callable(self):
```
Comment (gante) [inline tests/generation/test_utils.py:5047]: good test!

File: src/transformers/generation/utils.py (hunk 5 above)
Comment (manueldeprada) [reply src/transformers/generation/utils.py]: done!

File: src/transformers/generation/utils.py [hunk 10]
```diff
@@ -2227,11 +2227,16 @@ def generate(
...
+            custom_generate (`str` or `Callable`, *optional*):
+                One of the following:
+
+                - `str` (Hugging Face Hub repository name): runs the custom `generate` function defined at
+                  `custom_generate/generate.py` in that repository instead of the standard `generate` method. The
+                  repository fully replaces the generation logic, and the return type may differ.
+                - `str` (local repository path): same as above but from a local path, `trust_remote_code` not required.
+                - `Callable`: `generate` will perform the usual preparation steps, then call the provided callable to
+                  run the decoding loop. For more information, see
+                [Reusing generate's preparation steps by passing a callable](https://huggingface.co/docs/transformers/en/generation_strategies#reusing-generate-s-preparation-steps-by-passing-a-callable).
```
Comment (manueldeprada) [inline src/transformers/generation/utils.py]:
```suggestion
            custom_generate (`str` or `Callable`, *optional*):
                One of the following:
                - `str` (Hugging Face Hub repository name): runs the custom `generate` function defined at
                  `custom_generate/generate.py` in that repository instead of the standard `generate` method. The
                  repository fully replaces the generation logic, and the return type may differ.
                - `str` (local repository path): same as above but from a local path, `trust_remote_code` not required.
                - `Callable`: `generate` will perform the usual preparation steps, then call the provided callable to
                  run the decoding loop. For more information, see
                [the docs](https://huggingface.co/docs/transformers/en/generation_strategies#reusing-generate-s-preparation-steps-by-passing-a-callable).
```

File: src/transformers/generation/utils.py [hunk 11]
```diff
@@ -2235,11 +2235,15 @@ def generate(
...
-                repository, and the return type may be different from the standard `generate` method.
+            custom_generate (`str` or `Callable`, *optional*):
+                One of the following:
+                - `str` (Hugging Face Hub repository name): runs the custom `generate` function defined at
+                  `custom_generate/generate.py` in that repository instead of the standard `generate` method. The
+                  repository fully replaces the generation logic, and the return type may differ.
+                - `str` (local repository path): same as above but from a local path, `trust_remote_code` not required.
+                - `Callable`: `generate` will perform the usual preparation steps, then call the provided callable to
+                  run the decoding loop. For more information, see
+                [the docs](https://huggingface.co/docs/transformers/en/generation_strategies#reusing-generate-s-preparation-steps-by-passing-a-callable).
```
Comment (manueldeprada) [inline src/transformers/generation/utils.py]:
```suggestion
                [the docs](../../generation_strategies#reusing-generate-s-preparation-steps-by-passing-a-callable).
```

File: src/transformers/generation/utils.py [hunk 12]
```diff
@@ -2235,11 +2235,15 @@ def generate(
...
-                repository, and the return type may be different from the standard `generate` method.
+            custom_generate (`str` or `Callable`, *optional*):
+                One of the following:
+                - `str` (Hugging Face Hub repository name): runs the custom `generate` function defined at
+                  `custom_generate/generate.py` in that repository instead of the standard `generate` method. The
+                  repository fully replaces the generation logic, and the return type may differ.
+                - `str` (local repository path): same as above but from a local path, `trust_remote_code` not required.
+                - `Callable`: `generate` will perform the usual preparation steps, then call the provided callable to
+                  run the decoding loop. For more information, see
+                [the docs](../../generation_strategies#reusing-generate-s-preparation-steps-by-passing-a-callable).
```
Comment (manueldeprada) [inline src/transformers/generation/utils.py]:
```suggestion
                [the docs](../../generation_strategies#reusing-generate-s-input-preparation).
```

File: src/transformers/generation/utils.py [hunk 13]
```diff
@@ -2235,11 +2235,15 @@ def generate(
...
-                function defined in that reposity's `custom_generate/generate.py` file will be executed instead of the
-                standard `generate` method. Note that the logic is for generation is entirely defined in that
-                repository, and the return type may be different from the standard `generate` method.
+            custom_generate (`str` or `Callable`, *optional*):
+                One of the following:
+                - `str` (Hugging Face Hub repository name): runs the custom `generate` function defined at
+                  `custom_generate/generate.py` in that repository instead of the standard `generate` method. The
+                  repository fully replaces the generation logic, and the return type may differ.
+                - `str` (local repository path): same as above but from a local path, `trust_remote_code` not required.
+                - `Callable`: `generate` will perform the usual preparation steps, then call the provided callable to
```
Comment (gante) [inline src/transformers/generation/utils.py]:
```suggestion
                - `Callable`: `generate` will perform the usual input preparation steps, then call the provided callable to
```

File: src/transformers/generation/utils.py [hunk 14]
```diff
@@ -2235,11 +2235,15 @@ def generate(
...
+                  run the decoding loop. For more information, see
+                [the docs](../../generation_strategies#reusing-generate-s-input-preparation).
```
Comment (gante) [inline src/transformers/generation/utils.py]:
```suggestion
                  run the decoding loop. 
                For more information, see [the docs](../../generation_strategies#custom-generation-methods).
```
Sorry, I should have been more clear in the original comment: I wanted to suggest a link to the entry point of the `custom_generate` docs (as opposed to the callable-specific subsection). From this link, users can then figure out how to best use the flag in all cases, with examples.

**Review by gante — APPROVED** [review]
LGTM, thank you for working on it 💛
